src/key.rs

```diff
@@ -199,7 +199,34 @@ impl SecretKey {
     }
 }
 
-serde_impl!(SecretKey, constants::SECRET_KEY_SIZE);
+
+#[cfg(feature = "serde")]
+impl ::serde::Serialize for SecretKey {
```

Comment:
Seems like this is encouraging user to serialize their secrets. Do we really want to be doing that?

Reply:
As an alternative, we could provide a module so users have to explicitly opt-in to this by saying `#[serde(with = "secp256k1::serde::secret_key")]`. (This has the downside that we need to provide a module for `Option` and `Vec` as well if we want it to work out of the box with those containers.)

However, I am not sure if this makes things actually better? Some situations require secret keys to be serialized (a backup functionality for example). Additionally, `SecretKey` also implements `Debug` and `Display` which are basically the same.

I am not convinced that we should be introducing a policy on this abstraction level. Whether or not serializing a secret key is dangerous depends completely on the application.

Reply:
I'll just note that this is already the case: https://docs.rs/secp256k1/0.20.0/secp256k1/key/struct.SecretKey.html#impl-Serialize

Reply:
(At least as long as we don't have HD derivation natively in this library), serializing secret keys is essential. Keys need to be stored.

In one case, the caller has the secret key in serialized form already and just passes it to the library. Then letting the user serialize keys won't hurt either.

In the other case, the caller relies on the library to generate (or tweak) keys. Then it's essential to serialize them for storage.